R/request_functions.R

```diff
@@ -266,3 +261,86 @@ get_gbif_points = function(key) {
   print(res[1,2])
   return(res)
 }
+
+get_random_powo = function(){
+  
+  powo_fam_results <- tibble(
+    family=NA_character_,
+    accepted=NA_character_
+  )
+  
+  # use name full name to search API  
+  full_url =  paste("http://plantsoftheworldonline.org/api/1/search?f=accepted_names%2%2Cfamily_ff")
+  
+  # encode
+  full_url = utils::URLencode(full_url)
+  
+  # get raw json data
+  raw_data <- readLines(full_url, warn = "F", encoding = "UTF-8")
+  
+  # organise
+  rd = fromJSON(raw_data)
+  
+  # make data frame
+  results = rd$results
+  
+  #rd$totalResults
+  
+  # get IPNI ID
+  #results = mutate(results, IPNI_ID=str_extract(url, "(?<=names\\:)[\\d\\-]+$"))
+  
+  # only include these fields - you don't want synonym of
+  fam_results = select(results, colnames(powo_fam_results))
+  fam_results = subset(fam_results, accepted == "TRUE")
+  fam_results = subset(fam_results, select = family)
+  
+  # random page
+  #rand_page = sample(rd$totalPages,1)
+  
+  rand_page = sample(19,1)
+  
+  # run through cursors for n random pages
+  for (n in 1:rand_page){
+    
+    cursor = rd$cursor
+    full_url =  paste0("http://plantsoftheworldonline.org/api/1/search?cursor=",cursor)
+    raw_data <- readLines(full_url, warn = "F", encoding = "UTF-8")
+    rd = fromJSON(raw_data)
+  }
+
+  results = rd$results
+  #results = mutate(results, IPNI_ID=str_extract(url, "(?<=names\\:)[\\d\\-]+$"))
+  fam_results = select(results, family)
+    
+  # get random species from list
+  rand_fam = fam_results[sample(nrow(fam_results),1), ]
+  
+  # now get random species from that family
+  
+  #http://www.plantsoftheworldonline.org/?f=accepted_names%2Cspecies_f&q=Opiliaceae
+  
+  full_url = paste0("http://plantsoftheworldonline.org/api/1/search?f=accepted_names%2Cspecies_f&q=", rand_fam)
+  raw_data <- readLines(full_url, warn = "F", encoding = "UTF-8")
+  rd = fromJSON(raw_data)
+  
+  #rand_page = sample(50,1)
+  
+  #for (n in 1:rand_page){
+    
+  #  cursor = rd$cursor
+  #  full_url =  paste0("http://plantsoftheworldonline.org/api/1/search?cursor=",cursor)
```

Comment:
I think each time you need to use the whole query, including the filters and the random family name, but update the cursor. That should iterate through the page results.

Reply:
ok, makes sense.